Automattic/woocommerce-payments
# Consolidate gateways implementation - step I

This PR is the first step of https://github.com/Automattic/woocommerce-payments/issues/7782

#### Changes proposed in this Pull Request
The goal of https://github.com/Automattic/woocommerce-payments/issues/7782 is to
* merge three gateways (`WC_Payment_Gateway_WCPay`, `UPE_Payment_Gateway` and `UPE_Split_Payment_Gateway`) into one
* merge unit test suites following the previous changes

Merging three gateways in one PR is a challenging diff to analyse. Nothing stands in the way for us to perform this two-step merge separately: first merge split upe gateway into upe gateway, secondly merge upe gateway into classic gateway. This PR reflects the scope of `first merge split upe gateway into upe gateway`. Tests were partially cleaned up, but the majority of changes in the tests will land as part of the second PR. 

Looking high-level, this PR
* moves the logic from split UPE gateway into UPE gateway, sometimes by replacing the UPE implementation with the split UPE implementation because the former is not relevant anymore, sometimes by removing the non-used function(s)
* replaces all the occurrences of split UPE across the code, since UPE gateway is now capable of performing all the tasks
* removes split UPE gateway class
* removes some tests for the UPE gateway, which are now supposed to be only covered by the split UPE tests

It's up to you which review approach to choose, but my suggestion would be to split the code review into two steps
1. compare `class-upe-payment-gateway.php` and `class-upe-split-payment-gateway.php` classes in your IDE and verify that the new code in the `class-upe-payment-gateway.php` is applied from `class-upe-split-payment-gateway.php` in the expected way. In other words, understand where all the latest changes in the UPE gateway are coming from and why. There are also some removals (e.g. `init_setup_intent_ajax`) where we want to check if there are no risks of removing those.
2. review all the other files/changes in the diff. Those are usually name-change updates or some less complex operations.









#### Testing instructions





* Perform a purchase (any payment method, any checkout page type), navigate to Stripe dashboard -> Payment details and confirm that `gateway_type: split_upe_with_deferred_intent_creation` in gone from **Metadata**, but that other metadata fields are still there like e.g. `fraud_prevention_data_shopper_ua_hash`, `client_version` or `order_id`.
* Test the plugin update. Create a new JN site, set up WooPayments, navigate to Payments -> Settings -> Advanced settings and confirm Debug mode is enabled. Build zip using [this workflow](https://github.com/Automattic/woocommerce-payments/actions/workflows/build-zip-and-run-smoke-tests.yml) from `cleanup/upe-gateways` (feel free to skip the tests to save some time), update the plugin in your JN to the one from the built zip and confirm that there are no errors during the update. This test is to avoid a situation similar to this: p1702991804354289-slack-C011ENB20Q1
* enable multiple payment methods and confirm every gateway is displayed as expected on blocks & shortcode checkouts
* when paying with card & one other random gateway, confirm that `Payment method` field on the _order received_ page has the right payment method name
* for physical product purchases, ensure that `mandate_data` is attached to the `POST /v1/payment_intents` request when Stripe Link is enabled (request to be checked in Stripe dashboard)
* confirm that save payment method checkbox appears for card & physical products, and doesn't appear for any other non-reusable PM like e.g. giropay, Bancontact.
* Enable WooPay and 
    * Perform the below two checks, at least in Blocks but if possible - also in shortcode
    * Pay with a card & any additional payment method, and confirm that both payments are successful (use WooPayments checkout, not WooPay; we just want WooPay to be enabled at this point)
    * use WooPay to checkout and confirm the checkout is successful


-------------------


**Post merge**
